charmcraft/config.py

```diff
@@ -262,6 +262,35 @@ def validate_charm_type(cls, charm_type, values):
             raise ValueError("must be either 'charm' or 'bundle'")
         return charm_type
 
+    @classmethod
+    def expand_short_form_bases(cls, bases: List[Dict[str, Any]]) -> None:
+        """Expand short-form base configuration into long-form in-place."""
+        for index, base in enumerate(bases):
+            # Skip if already long-form. Account for common typos in case user
+            # intends to use long-form, but did so incorrectly (for better
+            # error message handling).
+            if (
+                "run-on" in base
+                or "run_on" in base
+                or "build-on" in base
+                or "build_on" in base
+            ):
```

Comment:
We also need to support this case:

```
bases:
  - build-on:
    - name: ubuntu
      channel: 20.04
      architectures: [<machine arch>]
```

Reply:
As in the 'run-on' is implied to match build-on?

Reply:
Yeap.

Reply:
Seems odd to support configuring the same thing three different ways out of the gate, but can do.

To clarify, the `run-on` implication is squarely the matching the *chosen* build-on that was used to create the build environment (as `build-on` is a list of options).  It may make sense to allow run-on to be None, in which case it will always be set to match the build environment.

How about we add this in a follow up PR?

Reply:
We've agreed that we will avoid this extra form for the time being as it provides no value.